src/Engine/ProtoCore/DSASM/Heap.cs

```diff
@@ -416,7 +421,29 @@ public StackValue AllocatePointer(int size, MetaData metadata)
         private StackValue AllocateStringInternal(string str, bool isConstant)
         {
             int index;
-            if (!stringTable.TryGetPointer(str, out index))
+            if (stringTable.TryGetPointer(str, out index)) {
+                // Any existing heap elements, marked as white, that are in the sweepSet and that are referenced during the sweep cycle will be marked as Black.
+                // This will ensure that no reachable data is mistakenly cleaned up.
+                bool isDuringGCPropagateCycle = gcState == GCState.Propagate;
+                bool isValidHeapIndex = index >= 0 && index < heapElements.Count;
+                if (isDuringGCPropagateCycle && isValidHeapIndex)
+                {
+                    var he = heapElements[index];
+                    if (he == null)
+                    {
+                        throw new NullReferenceException($"Null heap element found at index {index} during AllocateStringInternal");
```

Comment:
What happens when this exception is thrown? Is it a crash or is it caught somewhere? If it isn't expected in general, we add `Validity.Assert()` statements:
```
Validity.Assert(he != null, "Heap element at given index cannot be null");
```